tao_compiler/mlir/disc/transforms/disc_shape_optimization.cc

```diff
@@ -0,0 +1,1474 @@
+    if (failed(insertTieShapeOnBlock(&block))) return failure();
+  }
+  return success();
+}
+
+// convert:
+//   %shape = shape.shape_of %0 : tensor<?x?xf32> -> tensor<2xindex>
+// to:
+//   %d0 = tensor.dim %0, %c0 : tensor<?x?xf32>
+//   %d1 = tensor.dim %0, %c1 : tensor<?x?xf32>
+//   %shape = tensor.from_elements %d0, %d1 : tensor<2xindex>
+struct ExpandShapeOfOpPattern : public OpRewritePattern<shape::ShapeOfOp> {
+  using OpRewritePattern<shape::ShapeOfOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(shape::ShapeOfOp op,
+                                PatternRewriter& rewriter) const override {
+    auto ty = op.getResult().getType().dyn_cast<RankedTensorType>();
+    if (!ty || !ty.hasStaticShape() || !ty.getElementType().isIndex())
+      return failure();
+
+    SmallVector<Value> dimSizes;
+    for (int dim = 0, rank = ty.getShape()[0]; dim < rank; ++dim)
+      dimSizes.push_back(
+          rewriter.create<tensor::DimOp>(op.getLoc(), op.getArg(), dim));
+
+    rewriter.replaceOpWithNewOp<tensor::FromElementsOp>(op, dimSizes);
+    return success();
+  }
+};
+
+/// Fold dim of an operation that implements the InferShapedTypeOpInterface
+template <typename OpTy>
+struct DimOfShapedTypeOpInterface : public OpRewritePattern<OpTy> {
+  using OpRewritePattern<OpTy>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(OpTy dimOp,
+                                PatternRewriter& rewriter) const override {
+    OpResult dimValue = dimOp.source().template dyn_cast<OpResult>();
+    if (!dimValue) return failure();
+    auto shapedTypeOp =
+        dyn_cast<InferShapedTypeOpInterface>(dimValue.getOwner());
+    if (!shapedTypeOp) return failure();
+
+    Optional<int64_t> dimIndex = dimOp.getConstantIndex();
+    if (!dimIndex) return failure();
+
+    SmallVector<Value> reifiedResultShapes;
+    if (failed(shapedTypeOp.reifyReturnTypeShapes(
+            rewriter, shapedTypeOp->getOperands(), reifiedResultShapes)))
+      return failure();
+
+    if (reifiedResultShapes.size() != shapedTypeOp->getNumResults())
+      return failure();
+
+    Value resultShape = reifiedResultShapes[dimValue.getResultNumber()];
+    auto resultShapeType = resultShape.getType().dyn_cast<RankedTensorType>();
+    if (!resultShapeType ||
+        (!resultShapeType.getElementType().isa<IndexType>() &&
+         !resultShapeType.getElementType().isa<IntegerType>()))
+      return failure();
+
+    Location loc = dimOp.getLoc();
+    Value newValue = rewriter.create<tensor::ExtractOp>(
+        loc, resultShape,
+        rewriter.createOrFold<arith::ConstantIndexOp>(loc, *dimIndex));
+    if (!newValue.getType().isa<IndexType>())
+      newValue = rewriter.create<arith::IndexCastOp>(
+          loc, rewriter.getIndexType(), newValue);
+
+    rewriter.replaceOp(dimOp, newValue);
+    return success();
+  }
+};
+
+// materialize shape computation IR by resolving tensor.dim + op having
+// interface ShapedTypeOpInterface: before
```

Comment:
We let `tensor.extract` bypass `tie_shape` op to make normal `canonicalization & cse` patterns work. This will  be helpful to simplify shape computation IR.

In our current design, `tensor.extract` op can only be used to shape tensor (e.g. those must be placed on CPU). And all shape tensors are supposed to have static shape (since DISC is a static rank compiler).

Ideally, we should not insert a tie_shape for shape tensors. However, some tensor are known to be "shape tensor" after we insert `tie_shape` ops. For example, some `tensor.extract` ops are inserted during we resolve tensor.dim + InferShapedTypeOpInterface.

Reply:
Done, thanks!